server/server.go

```diff
@@ -32,55 +62,170 @@ func StartHttpServer(opts *options.Options) error {
 	if err != nil {
-		return err
+		return &CheckFail{name: c.Name, reason: err.Error()}
 	}
 
 	defer conn.Close()
+	logger.Infof("Check SUCCESS: %s", c.Name)
+	return nil
+}
 
+func (c HttpCheck) ValidateCheck () error {
+	if c.Name == "" {
+		return &InvalidCheck{name: "http", key: "name"}
+	}
+	if c.Host == "" {
+		return &InvalidCheck{name: c.Name, key: "host"}
+	}
+	if c.Port == 0 {
+		return &InvalidCheck{name: c.Name, key: "port"}
+	}
+	return nil
+}
+
+func (c HttpCheck) DoCheck (opts *options.Options) error {
+	logger := opts.Logger
+	logger.Infof("Checking %s at %s:%d via HTTP...", c.Name, c.Host, c.Port)
+
+	timeout := time.Second * DEFAULT_CHECK_TIMEOUT
+	if c.Timeout != 0 {
+		// override default with user defined timeout
+		timeout = time.Second * time.Duration(c.Timeout)
+	}
+
+	client := http.Client{
+		Timeout: timeout,
+	}
+	resp, err := client.Get(fmt.Sprintf("http://%s:%d", c.Host, c.Port))
+	if err != nil {
+		return &CheckFail{name: c.Name, reason: err.Error()}
+	}
+
+	switch {
+	case len(c.SuccessStatusCodes) > 0:
+		if !contains(c.SuccessStatusCodes, resp.StatusCode) {
+			return &CheckFail{name: c.Name, reason: fmt.Sprintf("wanted one of %v, got %d", c.SuccessStatusCodes, resp.Status)}
+		}
+	case c.BodyRegex != "":
+		defer resp.Body.Close()
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return &CheckFail{name: c.Name, reason: "failed reading body"}
+		}
+
+		if !strings.Contains(string(body), c.BodyRegex) {
+			return &CheckFail{name: c.Name, reason: fmt.Sprintf("wanted %s in http body, got %s", c.BodyRegex, body)}
+		}
+	default:
+		// no success_codes or body_regex defined, only pass on 200
+		if resp.StatusCode != http.StatusOK {
+			return &CheckFail{name: c.Name, reason: fmt.Sprintf("wanted status code 200, got %d", resp.StatusCode)}
+		}
+	}
+
+	logger.Infof("Check SUCCESS: %s", c.Name)
+	return nil
+}
+
+// TODO: move into helpers
+func contains(s []int, e int) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
+	}
+	return false
+}
+
+func (c ScriptCheck) ValidateCheck () error {
+	if c.Name == "" {
+		return &InvalidCheck{name: "script", key: "name"}
+	}
+	if c.Script == "" {
+		return &InvalidCheck{name: c.Name, key: "script"}
+	}
+	return nil
+}
+
+func (c ScriptCheck) DoCheck (opts *options.Options) error {
+	logger := opts.Logger
+	logger.Infof("Checking %s at %s...", c.Name, c.Script)
+
+	timeout := time.Second * DEFAULT_CHECK_TIMEOUT
+	if c.Timeout != 0 {
+		// override default with user defined timeout
+		timeout = time.Second * time.Duration(c.Timeout)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	cmd := exec.CommandContext(ctx, c.Script)
+	_, err := cmd.Output()
+	if ctx.Err() == context.DeadlineExceeded {
+		return &CheckTimeout{name: c.Name, timeout: int(timeout)}
+	}
+	if err != nil {
+		return &CheckFail{name: c.Name, reason: "non-zero exit code"}
```

Comment:
We should log this error and return it to help with debugging!

Reply:
Is it bad form to log it like this: https://github.com/gruntwork-io/health-checker/pull/9/files#diff-91bbeda7eb98a7adc57b9e47e2cf5c2bR79

It currently returns then logs it in that function

Reply:
Yes, that's totally fine. However, you should then return the underlying `err` from this method (e.g., it could be an optional field in the `CheckFail` struct that is rendered in its `Error()` method) so it gets logged by `DoCheck`.

Reply:
I get it now - thanks again @brikis98 - working on making sure the error/logging loose ends are all tied up and getting those checks tested... I think I'll feel better about "production" ready after that and a final final final CR :P